### DISCUSSION TITLE
Best Way to Handle Auth Records in Framework

### DISCUSSION THREAD

**Discussion by hlanderdev** [pocketbase/pocketbase#888] [Q&A] [answered]
First off, I am really enjoying the changes in [v0.8.0](https://github.com/pocketbase/pocketbase/releases/tag/v0.8.0-rc1), but I just had one question about how you envision Auth Records being handled when using Pocketbase as a framework. My first instinct when starting to use Pocketbase as a framework before the new release was to write structs with embedded models.Record structs and writing helper methods to set and get specific fields in the data map. Then, I learned from #734 about creating your own model structs by embedding the models.BaseModel struct and implementing the models.Model interface by defining a TableName() method. This provides way more structure to my project, and is how I have been doing it.

In v0.8.0, however, if you try to implement an Auth Record this way, you lose a some functionality. One example is that you cannot generate new auth tokens with tokens.NewRecordAuthToken(), as that specifically requires a record struct as a argument. The ability to do this is specifically helpful in running endpoint tests that need an auth header. Another small thing is that it becomes awkward with the built-in middleware that stores the Auth Record in the request context. You end up having to either do another DB query to get the user data in your desired model struct, or you need to write your own helper functions to convert the Auth Record to your desired model struct.

Maybe I'm jumping the gun and all of this will be explained in the new docs when they are released, but any clarity will be really appreciated. Thanks!

**Comment by ganigeorgiev** [comment, marked as answer]
Converting the fixed `models.User` to the dynamic `models.Record` is definitely not perfect and comes with some limitations.

`tokens.NewRecord*Token()` functions works only with `models.Record` and I'm not sure if we can change this. If you want  to generate tokens from your custom model struct you'll have to create your own helpers. The `tokens.*` helpers doesn't do much and calls `security.NewToken` with the extracted fields from the arguments.

Eventually we can create builtin generic helpers but Go is very restrictive and I'm not sure how this will look. Replacing everything with interfaces is also not ideal because in this case you'll have to define getter methods for the used fields and it will be too verbose. I'll have to think a little more on this but feel free to let me know if you have any suggestions.

**Reply by hlanderdev** [reply]
It looks like `models.Record` is already in such a way that it could implement a `AuthModel` interface, since `models.Record` already has getters and setters anyway. It might not be perfect, but you could follow a similar approach to the `models.Model` interface and the `models.BaseModel` struct:

```
type AuthModel interface {
    Model // embed the base Model interface
    Username() string
    SetUsername(name string) error
    // etc...
}

type BaseAuthModel struct {
    BaseModel // embed the BaseModel struct
    Username string `db:"username" json:"username"`
    // etc...
}

func (m *BaseAuthModel) Username() string {
    return m.Username
}

// etc...
```

**Reply by ganigeorgiev** [reply]
There are 2 issues with the interface approach:

1. We may need access to the collection instance (eg. `record.Collection()`) but let's assume that we may change the internals and will resolve the collection from a cache based on the identifier (in this case it will be the table name).

2. Adding an AuthModel interface will be too verbose if we require the user to implement all auth related methods:
   ```go
   func (m *Record) Username() string
   func (m *Record) SetUsername(username string) error
   func (m *Record) Email() string
   func (m *Record) SetEmail(email string) error
   func (m *Record) EmailVisibility() bool
   func (m *Record) SetEmailVisibility(visible bool) error
   func (m *Record) Verified() bool
   func (m *Record) SetVerified(verified bool) error
   func (m *Record) TokenKey() string
   func (m *Record) SetTokenKey(key string) error
   func (m *Record) RefreshTokenKey() error
   func (m *Record) LastResetSentAt() types.DateTime
   func (m *Record) SetLastResetSentAt(dateTime types.DateTime) error
   func (m *Record) LastVerificationSentAt() types.DateTime
   func (m *Record) SetLastVerificationSentAt(dateTime types.DateTime) error
   func (m *Record) ValidatePassword(password string) bool
   func (m *Record) SetPassword(password string) error
   ```
   While we can restrict the token method interface to a limited set of getter fields, I think we will end up with a lot of slightly similar interfaces and it will become too ambiguous to use an navigate in the codebase.
   
At the end if you are going with the troubles of defining a getter/setter for each of your fields, it may be better to just embed the `Record` model.

```go
type Example struct {
  *models.Record
}

func (e *Example) Name() string {
  return e.Record.GetString("name")
}

... 

var example Example

tokens.NewRecordAuthToken(app, example.Record)
```

**Reply by ganigeorgiev** [reply]
Actually, on second read I think I understand what do you mean. It may be doable without even requiring to double map the auth fields, I'll give it a try.

**Reply by hlanderdev** [reply]
I see, but in my example above all of the auth specific methods would be pre-defined in the `models.BaseAuthModel` struct. Then similar to `models.BaseModel` the user could just embed it then only have to defined the `TableName()` method:

```
var _ models.AuthModel = (*MyAuthModel)(nil)

type MyAuthModel struct {
    models.BaseAuthModel
    MyField string `db:"my_field" json:"my_field"`
}

func (m *MyAuthModel) TableName() string {
    return "my_auth_collection"
}
```

Or am I missing a key point here?

Edit: I just saw the above comment.

**Reply by ganigeorgiev** [reply]
@highlanderdev 

I've played a little with the suggestion, but I don't think we can implement it properly because the auth models have conditional serialization and if we provide a default `json.Marshaler` implementation in the embedded `BaseAuthModel` then the custom model struct fields will no longer be seriazable unless the users provide their own custom `json.Marshaler`. For example:

```go
package main

import (
  "encoding/json"
  "fmt"
)

type BaseAuthModel struct {
  Email string `db:"email" json:"email"`
}

func (m BaseAuthModel) MarshalJSON() ([]byte, error) {
  return json.Marshal(map[string]any{
    "base": 123,
  })
}

// ---

type MyModel struct {
  BaseAuthModel

  Name string `json:"name"`
}

func main() {
  raw, _ := json.Marshal(MyModel{})

  // Output: {"base": 123}
  fmt.Println(string(raw))
}
```

One solution would be to just not have a base `MarshalJSON` and leave the users to decide whether they want to serialize the special fields (like email, tokenKey, etc.) but this probably would result in accidental sensitive data leaks because it is possible in the token helpers for example to add the exported record model or part of it as jwt payload in the future or by just writing the custom auth model to the response body in some custom endpoint.

The auth records have domain specific logic and I'm not sure that the use cases where you would want to pass a custom auth record to an internal record related function justifies the eventual complexity down the road. Even if we implement the above, it seems that it will be usable at the moment only for the `token.NewRecord*Token` helpers.

Let's wait and see what other use cases may arise before adding a new interface because it will be difficult to remove later.

**Reply by hlanderdev** [reply]
Ah, I see. I didn't think of this consequence. You're right. It's probably best to wait. then.
